### PR TITLE
refactor: free list page ref & mut

### PR DESCRIPTION
**Pull request by pepyakin**
Adds types that encapsulate manipulation of a free list page.

### CODE REVIEW

**Comment by pepyakin** [comment]
* **#399** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/399?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @pepyakin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: nomt/src/beatree/allocator/free_list.rs [hunk 1]
```diff
@@ -369,15 +348,55 @@ fn decode_free_list_page(page: FatPage) -> (PageNumber, Vec<PageNumber>) {
...
+impl<'a> FreeListPageMut<'a> {
+    fn set_prev_pn(&mut self, pn: PageNumber) {
+        self.0[0..4].copy_from_slice(&pn.0.to_le_bytes());
+    }
+
+    fn set_item_count(&mut self, count: u16) {
+        self.0[4..6].copy_from_slice(&count.to_le_bytes());
+    }
+
+    fn set_item(&mut self, i: usize, pn: PageNumber) {
```
Comment (deuszx) [inline nomt/src/beatree/allocator/free_list.rs:396]: This will throw if you try to access index outside of the underlying array - is that OK?

Comment (pepyakin) [reply nomt/src/beatree/allocator/free_list.rs:396]: should be fine. Less for for a public API, but this is not meant to be one. The whole change is just trying to improve legibility of the code.

**Review by rphmeier — APPROVED** [review]

File: nomt/src/beatree/allocator/free_list.rs (hunk 1 above)
Comment (rphmeier) [reply nomt/src/beatree/allocator/free_list.rs:396]: Yeah. In a public API it's worth at least documenting, but since it's private it should be fine.

**Pull request merged**